### ISSUE TITLE
[BUG] Docker build fails on unable to satisfy requirement for sphinx 2.2.1

### ISSUE THREAD

**Issue by jimthompson5802** [mlflow/mlflow#2101]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in MLflow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: MacOS 10.15.1
- **MLflow installed from (source or binary)**: N/A
- **MLflow version (run ``mlflow --version``)**:  1.4.1.dev0
- **Python version**: N/A
- **npm version, if running the dev UI**: N/A
- **Exact command to reproduce**:

Issue command `docker build -t mlflow_dev .`

### Describe the problem
docker build fails with this error:
```
8 26.22 Collecting sphinx==2.2.1 (from -r dev-requirements.txt (line 2))
#8 26.47   ERROR: Could not find a version that satisfies the requirement sphinx==2.2.1 (from -r dev-requirements.txt (line 2)) (from versions: 0.1.61611, 0.1.61798, 0.1.61843, 0.1.61945, 0.1.61950, 0.2, 0.3, 0.4, 0.4.1, 0.4.2, 0.4.3, 0.5, 0.5.1, 0.5.2b1, 0.5.2, 0.6b1, 0.6, 0.6.1, 0.6.2, 0.6.3, 0.6.4, 0.6.5, 0.6.6, 0.6.7, 1.0b1, 1.0b2, 1.0, 1.0.1, 1.0.2, 1.0.3, 1.0.4, 1.0.5, 1.0.6, 1.0.7, 1.0.8, 1.1, 1.1.1, 1.1.2, 1.1.3, 1.2b1, 1.2b2, 1.2b3, 1.2, 1.2.1, 1.2.2, 1.2.3, 1.3b1, 1.3b2, 1.3b3, 1.3, 1.3.1, 1.3.2, 1.3.3, 1.3.4, 1.3.5, 1.3.6, 1.4a1, 1.4b1, 1.4, 1.4.1, 1.4.2, 1.4.3, 1.4.4, 1.4.5, 1.4.6, 1.4.7, 1.4.8, 1.4.9, 1.5a1, 1.5a2, 1.5b1, 1.5, 1.5.1, 1.5.2, 1.5.3, 1.5.4, 1.5.5, 1.5.6, 1.6b1, 1.6b2, 1.6b3, 1.6.1, 1.6.2, 1.6.3, 1.6.4, 1.6.5, 1.6.6, 1.6.7, 1.7.0b1, 1.7.0b2, 1.7.0, 1.7.1, 1.7.2, 1.7.3, 1.7.4, 1.7.5, 1.7.6, 1.7.7, 1.7.8, 1.7.9, 1.8.0b1, 1.8.0, 1.8.1, 1.8.2, 1.8.3, 1.8.4, 1.8.5)
#8 26.47 ERROR: No matching distribution found for sphinx==2.2.1 (from -r dev-requirements.txt (line 2))
------
```

### Code to reproduce issue
Use Dockerfile in the master branch

### Other info / logs
Here is the full docker build error with error.
[docker_build_error_2019_11_14.txt](https://github.com/mlflow/mlflow/files/3849660/docker_build_error_2019_11_14.txt)

The package sphinx 2.2.1 now requires Python 3.  The base image used in the Dockerfile is `continuumio/miniconda`, which provides a Python 2.x environment.  If it is reasonable, I can submit a PR to that changes the base image to miniconda3.

**Comment by jimthompson5802** [comment]
After changing the base image to miniconda3, the docker build fails with
```
#8 58.51 ERROR: Could not open requirements file: [Errno 2] No such file or directory: 'travis/small-requirements.txt'
```
From what I can tell this is the result of placing `travis` in `.dockerignore`.

At this point, I'm not sure how to proceed without inadvertently affecting some other aspect of mlflow design objective.

**Comment by jimthompson5802** [comment]
To address the `No such file or directory: 'travis/small-requirements.txt'` error, modified `.dockerignore` as follows:
```

travis
# required for Docker build
!travis/small-requirements.txt
!travis/large-requirements.txt
!travis/lint-requirements.txt

tests
# required for Docker build
!tests/resources/mlflow-test-plugin/
```
This allows docker build to progress past the`No such file` error.  However, it now fails during the install of onnx: 
```
#8 311.6   Stored in directory: /root/.cache/pip/wheels/76/73/b8/8639930fbc56e9f26fac210f5e65dbe05a0c39396ddaf4a685
#8 312.6   Building wheel for onnx (setup.py): started
#8 314.4   Building wheel for onnx (setup.py): finished with status 'error'
#8 314.4   ERROR: Command errored out with exit status 1:
#8 314.4    command: /opt/conda/bin/python -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-ij9pgjn0/onnx/setup.py'"'"'; __file__='"'"'/tmp/pip-install-ij9pgjn0/onnx/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-5eeirx3u --python-tag cp37
#8 314.4        cwd: /tmp/pip-install-ij9pgjn0/onnx/
#8 314.4   Complete output (64 lines):
#8 314.4   fatal: not a git repository (or any of the parent directories): .git
#8 314.4   running bdist_wheel
#8 314.4   running build
#8 314.4   running build_py
#8 314.4   running create_version
#8 314.4   running cmake_build
#8 314.4   -- Build type not set - defaulting to Release
#8 314.4   -- The C compiler identification is GNU 8.3.0
#8 314.4   -- The CXX compiler identification is GNU 8.3.0
#8 314.4   -- Check for working C compiler: /usr/bin/cc
#8 314.4   -- Check for working C compiler: /usr/bin/cc -- works
#8 314.4   -- Detecting C compiler ABI info
#8 314.4   -- Detecting C compiler ABI info - done
#8 314.4   -- Detecting C compile features
#8 314.4   -- Detecting C compile features - done
#8 314.4   -- Check for working CXX compiler: /usr/bin/c++
#8 314.4   -- Check for working CXX compiler: /usr/bin/c++ -- works
#8 314.4   -- Detecting CXX compiler ABI info
#8 314.4   -- Detecting CXX compiler ABI info - done
#8 314.4   -- Detecting CXX compile features
#8 314.4   -- Detecting CXX compile features - done
#8 314.4   CMake Error at CMakeLists.txt:217 (message):
#8 314.4     Protobuf compiler not found
#8 314.4   Call Stack (most recent call first):
#8 314.4     CMakeLists.txt:248 (relative_protobuf_generate_cpp)
#8 314.4
#8 314.4
#8 314.4   -- Configuring incomplete, errors occurred!
#8 314.4   See also "/tmp/pip-install-ij9pgjn0/onnx/.setuptools-cmake-build/CMakeFiles/CMakeOutput.log".
#8 314.4   Traceback (most recent call last):
#8 314.4     File "<string>", line 1, in <module>
#8 314.4     File "/tmp/pip-install-ij9pgjn0/onnx/setup.py", line 328, in <module>
#8 314.4       'backend-test-tools = onnx.backend.test.cmd_tools:main',
#8 314.4     File "/opt/conda/lib/python3.7/site-packages/setuptools/__init__.py", line 145, in setup
#8 314.4       return distutils.core.setup(**attrs)
#8 314.4     File "/opt/conda/lib/python3.7/distutils/core.py", line 148, in setup
#8 314.4       dist.run_commands()
#8 314.4     File "/opt/conda/lib/python3.7/distutils/dist.py", line 966, in run_commands
#8 314.4       self.run_command(cmd)
#8 314.4     File "/opt/conda/lib/python3.7/distutils/dist.py", line 985, in run_command
#8 314.4       cmd_obj.run()
#8 314.4     File "/opt/conda/lib/python3.7/site-packages/wheel/bdist_wheel.py", line 192, in run
#8 314.4       self.run_command('build')
#8 314.4     File "/opt/conda/lib/python3.7/distutils/cmd.py", line 313, in run_command
#8 314.4       self.distribution.run_command(command)
#8 314.4     File "/opt/conda/lib/python3.7/distutils/dist.py", line 985, in run_command
#8 314.4       cmd_obj.run()
#8 314.4     File "/opt/conda/lib/python3.7/distutils/command/build.py", line 135, in run
#8 314.4       self.run_command(cmd_name)
#8 314.4     File "/opt/conda/lib/python3.7/distutils/cmd.py", line 313, in run_command
#8 314.4       self.distribution.run_command(command)
#8 314.4     File "/opt/conda/lib/python3.7/distutils/dist.py", line 985, in run_command
#8 314.4       cmd_obj.run()
#8 314.4     File "/tmp/pip-install-ij9pgjn0/onnx/setup.py", line 203, in run
#8 314.4       self.run_command('cmake_build')
#8 314.4     File "/opt/conda/lib/python3.7/distutils/cmd.py", line 313, in run_command
#8 314.4       self.distribution.run_command(command)
#8 314.4     File "/opt/conda/lib/python3.7/distutils/dist.py", line 985, in run_command
#8 314.4       cmd_obj.run()
#8 314.4     File "/tmp/pip-install-ij9pgjn0/onnx/setup.py", line 190, in run
#8 314.4       subprocess.check_call(cmake_args)
#8 314.4     File "/opt/conda/lib/python3.7/subprocess.py", line 347, in check_call
#8 314.4       raise CalledProcessError(retcode, cmd)
#8 314.4   subprocess.CalledProcessError: Command '['/usr/bin/cmake', '-DPYTHON_INCLUDE_DIR=/opt/conda/include/python3.7m', '-DPYTHON_EXECUTABLE=/opt/conda/bin/python', '-DBUILD_ONNX_PYTHON=ON', '-DCMAKE_EXPORT_COMPILE_COMMANDS=ON', '-DONNX_NAMESPACE=onnx', '-DPY_EXT_SUFFIX=.cpython-37m-x86_64-linux-gnu.so', '/tmp/pip-install-ij9pgjn0/onnx']' returned non-zero exit status 1.
#8 314.4   ----------------------------------------
```
This error message appears several times during the onnx build.

**Comment by jimthompson5802** [comment]
More issues with docker build.  Assuming I get past the above issues, a failure occurs in the install for openjdk-8-jre-headless:
```
#8 374.9 Package openjdk-8-jre-headless is not available, but is referred to by another package.
#8 374.9 This may mean that the package is missing, has been obsoleted, or
#8 374.9 is only available from another source
#8 374.9
#8 375.0 E: Package 'openjdk-8-jre-headless' has no installation candidate
```
In researching this error, I come across [this issue from stackoverlfow](https://stackoverflow.com/questions/57593809/dockerfile-installation-package-openjdk-8-jre-headless-has-no-installatio), which implies the error only occurs during a docker build.

**Comment by jimthompson5802** [comment]
Finally have a  working Dockerfile that can be build an image.  Summary of changes:
* changed base image to continuumio/miniconda3 because at least one pre-prequiste package (sphinx 2.2.1) requires Python 3.
* Added install for system packages `cmake`, `protobuf-compiler` to support install of `onnx` related packages
* Added install for system package `curl` to support install of npm
* Migrated from `openjdk-8-jre-headless` to `openjdk-11-jre-headless` because Debian 10 no longer supports `openjdk-8`.

I'll submit PR to implement the above.

**Comment by jimthompson5802** [comment]
@juntai-zheng   You helped me with a recent PR.  

I'm hoping you are able to provide guidance on this issue.    From my testing there appears to be several issues with the current mlflow Dockerfile.  I'm looking for feedback on whether the proposed fixes in PR #2107 are a reasonable solution.

**Comment by juntai-zheng** [comment]
Hi @jimthompson5802, thanks for taking a look at this! I'll rerun the Travis tests and take a look at your PR - seems like a good fix. Do you think you could write a test for this so we don't run into this issue in the future?

**Comment by jimthompson5802** [comment]
@juntai-zheng   Thank you for the response.

AFAIK, there is nothing in the current Travis test that validates the `docker build`.   The Travis tests basically test only Python/R related code.  This is a reason that issues like #1250 and #1394.

When you write, "Do you think you could write a test for this so we don't run into this issue in the future?"   I'm assuming you mean to test **building the docker image**.  If this is the case, I don't think the `pytest` approach will meet that need.  

In researching this, I found [this write-up](https://bencane.com/2016/01/11/using-travis-ci-to-test-docker-builds/) on testing  building a docker image using Travis.  If this is what you had in mind, I can take  stab at doing the test but it will require modifying `.travis.yml`.  Since this is a key foundation to mlflow's CI/CD pipeline, I want to confirm that it is you are looking for me to do.

**Comment by juntai-zheng** [comment]
Yup, I think editing the `travis.yml` file would be the right move. You should probably add the test to the nightly stage as a new Docker section, as the test would probably be pretty heavy and not needed for every PR. You will have to change [line 18](https://github.com/mlflow/mlflow/blob/master/.travis.yml#L18) to check that the Dockerfile changed similar to the check for seeing if anything related to your `examples` tests has changed. 

Travis can be pretty confusing, so if you have any questions on how to approach this feel free to ask!

**Comment by jimthompson5802** [comment]
OK...I'll work on implementing the test for building the docker image.

**Comment by jimthompson5802** [comment]
Fixed by PR #2107
